code/iaas/model/src/main/java/io/cattle/platform/core/cleanup/TableCleanup.java

```diff
@@ -249,81 +269,115 @@ private void cleanupLabelTables(Date cutoff) {
+                Table<?> referencingTable = key.getTable();
+                Field<Long> foreignKeyField = (Field<Long>) key.getFields().get(0);
+                idsReferencedQuery = create().selectDistinct(foreignKeyField).from(referencingTable);
+                for(Record1<Long> record : idsReferencedQuery.fetch()) {
+                    idsReferencedInOtherTables.add(record.value1());
                 }
             }
 
-            while (toDelete.size() > 0) {
+            while (idsToDeleteResult.size() > 0) {
+                // First organize rows into sets that can and cannot be deleted
                 HashSet<Long> idsToDelete = new HashSet<Long>();
-                for (Record1<Long> record : toDelete) {
-                    if (idsToFilter.contains(record.value1())) {
-                        idsToFix.add(record.value1());
-                        idsToFilter.remove(record.value1());
+                Long id = null;
+                for (Record1<Long> record : idsToDeleteResult) {
+                    id = record.value1();
+                    if (idsReferencedInOtherTables.contains(id)) {
+                        idsThatCantBeDeleted.add(id);
+                        idsReferencedInOtherTables.remove(id);
                     }
-                    if ((!idsToFix.contains(record.value1()))) {
-                        idsToDelete.add(record.value1());
+                    if (!idsThatCantBeDeleted.contains(id)) {
+                        idsToDelete.add(id);
                     }
                 }
-                if (idsToDelete.size() == 0) {
-                    break;
-                }
-                try {
-                    table.addRowsDeleted(create()
-                            .delete(table.table)
-                            .where(id.in(idsToDelete))
-                            .execute());
-
-                } catch (org.jooq.exception.DataAccessException e) {
-                    log.info(e.getMessage());
-                    break;
-                }
-                if (idsToFix.size() > 0) {
-                    ids = create().select(id).from(table.table).where(remove.lt(cutoffTime)).orderBy(id)
-                    .limit(QUERY_LIMIT_ROWS.getValue()).offset(idsToFix.size());
+                Long lastId = id;
+
+                // Delete rows that aren't referenced by other tables
+                if (idsToDelete.size() > 0) {
+                    try {
+                        int rowsDeleted = create()
+                                .delete(table.table)
+                                .where(idField.in(idsToDelete))
+                                .execute();
+                        table.addRowsDeleted(rowsDeleted);
+                        log.debug("Deleted {} unneeded rows from {} table", idsToDelete.size(), table.table);
+                    } catch (DataAccessException e) {
+                        log.info("Problem executing cleanup query: " + e.getMessage());
```

Comment:
Nit: should we log it in WARN?

Reply:
This was set to info before, i dont want to change the level